internal/testtypes/matchers.go

```diff
@@ -0,0 +1,275 @@
+func (m *Matcher) FailureMessage(other interface{}) (message string) {
+	return fmt.Sprintf("the following fields were expected to match but did not:\n%s\n%s", string(m.diff),
+		format.Message(other, "expected to match", m.original))
+}
+
+// NegatedFailureMessage returns a string comparing the full objects after an unexpected match has occurred.
+func (m *Matcher) NegatedFailureMessage(other interface{}) (message string) {
+	return format.Message(other, "not to match", m.original)
+}
+
+// calculateDiff applies the MatchOptions and identified the diff between the Matcher object and the passed object.
+func (m *Matcher) calculateDiff(other interface{}) ([]byte, error) {
+	// Convert the original and other objects to json.
+	originalJSON, err := json.Marshal(m.original)
+	if err != nil {
+		return nil, err
+	}
+
+	modifiedJSON, err := json.Marshal(other)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use a mergePatch to produce a diff between the two objects.
+	originalWithModifiedJSON, err := jsonpatch.MergePatch(originalJSON, modifiedJSON)
+	if err != nil {
+		return nil, err
+	}
+	rawDiff, err := jsonpatch.CreateMergePatch(originalJSON, originalWithModifiedJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	// Filter the diff according to the rules attached to the Matcher.
+	diff, err := filterDiff(rawDiff, m.options.allowPaths, m.options.ignorePaths)
+	if err != nil {
+		return nil, err
+	}
+	return diff, nil
+}
+
+// MatchOption describes an Option that can be applied to a Matcher.
+type MatchOption interface {
+	// ApplyToMatcher applies this configuration to the given MatchOption.
+	ApplyToMatcher(options *MatchOptions)
+}
+
+// MatchOptions holds the available types of MatchOptions that can be applied to a Matcher.
+type MatchOptions struct {
+	ignorePaths [][]string
+	allowPaths  [][]string
+}
+
+// ApplyOptions adds the passed MatchOptions to the MatchOptions struct.
+func (o *MatchOptions) ApplyOptions(opts []MatchOption) *MatchOptions {
+	for _, opt := range opts {
+		opt.ApplyToMatcher(o)
+	}
+	return o
+}
+
+// IgnorePaths instructs the Matcher to ignore given paths when computing a diff.
+type IgnorePaths [][]string
+
+// ApplyToMatcher applies this configuration to the given MatchOptions.
+func (i IgnorePaths) ApplyToMatcher(opts *MatchOptions) {
+	opts.ignorePaths = append(opts.ignorePaths, i...)
+}
+
+// AllowPaths instructs the Matcher to restrict its diff to the given paths. If empty the Matcher will look at all paths.
+type AllowPaths [][]string
+
+// ApplyToMatcher applies this configuration to the given MatchOptions.
+func (i AllowPaths) ApplyToMatcher(opts *MatchOptions) {
+	opts.allowPaths = append(opts.allowPaths, i...)
+}
+
+// filterDiff limits the diff to allowPaths if given and excludes ignorePaths if given. It returns the altered diff.
+func filterDiff(diff []byte, allowedPaths, ignorePaths [][]string) ([]byte, error) {
```

Comment:
```suggestion
func filterDiff(diff []byte, allowPaths, ignorePaths [][]string) ([]byte, error) {
```
nit (to keep it consistent with the field name)